Issue #1800 Add XWiki Support

dchenbecker opened the pull request:
This is very rudimentary for now. Based on DocuWiki/MediaWiki, but with less
support for special formatting because it's unclear what additional support
XWiki has for this.

SimplGy commented:
Way to get after it, Derek. I'd love to have an xwiki writer and hope this gets merged in!

marcbowes commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,243 @@
+   Stability   : alpha
+   Portability : portable
+
+Conversion of 'Pandoc' documents to DokuWiki markup.
```
Stale comment? ("DokuWiki")

dchenbecker replied:
Fixing.

mb21 commented:
Thanks!

What do you mean with "special formatting"?

Also, I assume XWiki merits its own writer and shouldn't be a variant of the mediawiki writer? (If there's a lot of code duplication, it could be done like e.g. `markdown_mmd`)

dchenbecker commented:
For special formatting I meant things like html tags. I do see now that there's support for inline element attributes (via the `(% <attr>="..." %)` tag, and I just found the section on escaping HTML tags directly, so I can start working on that. In terms of code duplication, this will look a lot like DocuWiki and MediaWiki, so if you think it's worth factoring that into a common set of functions I can take a stab at it, but it's going to be a bigger chunk of work and I would prefer to get something that works generally before iterating on improvements.

mb21 commented:
What I meant was to modify the `Mediawiki.hs` file to check for extensions (similarly to how the `Markdown.hs` checks whether `Ext_*` are defined. (The `blockToXWiki (Header level _ inlines) =` caught my eye because it's the same as the Mediawiki one.) But on a closer look, the two writers seem to be different enough, so I guess it makes more sense the way you've started it.

btw, the following is a good test case (check the generated `writer.xwiki` _very_ carefully before checking in):

    pandoc -s -f native test/testsuite.native -o test/writer.xwiki

mb21 commented:
P.S. maybe you want to use [Data.Text](https://www.stackage.org/haddock/lts-10.0/text-1.2.2.2/Data-Text.html) instead of `String`? Add `{-# LANGUAGE OverloadedStrings #-}` to the top, then it should be mostly the same except:

- `pack body` -> `body`
- `++` -> `<>`

Alternatively, especially if you have to indent a lot of things, [Text.Pandoc.Pretty](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Pretty.hs) may be a better choice? Lots of writers already use it...

dchenbecker commented:
I tried using OverloadedStrings and everything blew up. I really like Haskell, but Text/String is a perpetual source of confusion and frustration. If there's a strong reason for using it I'll take another stab at it.

dchenbecker commented:
OK, took another stab at String -> Text and got it working. I'm notsure what I'm doing wrong, but hoogle didn't return relevant results for "<>", so it took me a while to sort out what I actually needed (Data.Monoid). Kind of ironic that Monoids are one of my favorite FP concepts, but I couldn't figure out the syntax around them...

mb21 commented:
cool, btw. because `<>` is an infix function, you’d have to hoogle `(<>)`

dchenbecker commented:
OK, after way too long I have some time to work on this again. I've made some additional fixes and checked in a writer.xwiki file. The formatting is about 95% correct, but getting that last 5% is going to take some thinking and better understanding of the code, so I would like to at least ship what I have so far. 

If the changes look good, I need to resolve a small conflict. Would you prefer that I just merge and fix conflicts, or do you want all my work rebased/squashed into a single commit?

jgm commented:
I'd prefer rebase/squash.

Can you comment on what is in that last 5%?

Derek Chen-Becker <notifications@github.com> writes:

> OK, after way too long I have some time to work on this again. I've made some additional fixes and checked in a writer.xwiki file. The formatting is about 95% correct, but getting that last 5% is going to take some thinking and better understanding of the code, so I would like to at least ship what I have so far. 
>
> If the changes look good, I need to resolve a small conflict. Would you prefer that I just merge and fix conflicts, or do you want all my work rebased/squashed into a single commit?
>
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/jgm/pandoc/pull/4167#issuecomment-415451712

dchenbecker commented:
The last 5% currently includes these items:

* Custom bullet formatting
* Properly handling loose lists (numbering is currently reset)
* Footnotes

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+
+-- | Convert Pandoc to XWiki.
+writeXWiki :: PandocMonad m => WriterOptions -> Pandoc -> m Text
+writeXWiki _ (Pandoc _ blocks) =
```
The `in do` here, while perfectly legal, is not very idiomatic.  Usually one would do this instead:
```
writeXWiki _ (Pandoc _ blocks) = do
   let env = WriterState { listLevel = "" }
   body <- runReaderT (blockListToXWiki blocks) env
   return $ body
```

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+
+blockToXWiki Null = return ""
+
+-- TODO: handle this
```
What handling does this need?
If there is a way to mark internal anchors in xwiki, one could check the id attribute and set an anchor.

dchenbecker replied:
I've added this, but there doesn't seem to be a test case in testsuite.native where a Div has an id set (they're all ""):

```
9 matches for "Div" in buffer: testsuite.native
    231:,Div ("",[],[])
    234:,Div ("",[],[])
    235: [Div ("",[],[])
    236:  [Div ("",[],[])
    238: ,Div ("",[],[])
    253:,Div ("",[],[])
    260:,Div ("",[],[])
    261: [Div ("",[],[])
    262:  [Div ("",[],[])
```

dchenbecker replied:
OK, confirmed this is working with a quick `text.html` file:

```html
<div id="test">
This is div 3
</div>

<a href="#test">Link to test div</a>
```

produces:

```xwiki
{{ id name="test" /}}This is div 3
[[Link to test div>>#test]]
```

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+formRow :: PandocMonad m => [[Block]] -> XWikiReader m Text
+formRow row = do
+  cellStrings <- mapM (tableCellXWiki False) row
+  return $ Data.Text.unwords cellStrings
```
You may know this, but it's usual to do something like
```
import qualified Data.Text as Text
```
and then `Text.unwords` works.

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+
+inlineListToXWiki :: PandocMonad m => [Inline] -> XWikiReader m Text
+inlineListToXWiki lst =
+  fmap Data.Text.concat $ mapM inlineToXWiki lst
```
You may know this, but it's more usual to use the infix form:
```
Data.Text.concat <$> mapM inlineToXWiki lst
```
Since `Data.Text.concat` is monoid mconcat, you could also just do
```
mconcat <$> mapM inlineToXWiki lst
```

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+
+inlineToXWiki LineBreak = return "\n"
+
+-- FIXME: better understand this
```
We can't render a SoftBreak as a newline, because a newline is interpreted as a hard break in xwiki -- or so I gather from your other code.  Hence it becomes a space.

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+  return $ vcat contents <> if Text.null lev then "\n" else ""
+
+-- TODO: support more features
+blockToXWiki (Table _ _ _ headers rows') = do
```
Will this work even for complex table cells containing multiple blocks?
Might want to fall back to raw HTML for complex tables.
See other wiki writers.

dchenbecker replied:
XWiki doesn't support raw html, unfortunately. I found this out while trying to fix RawBlock (below). I can take a look at how DokuWiki does non-raw table handling and see if I can adapt it.

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+  contents <- inlineListToXWiki lst
+  return contents
+
+inlineToXWiki (Quoted SingleQuote lst) = do
```
These are for "smart quotes" - does xwiki interpret straight quotes as curly automatically?
If not, might want to use curly quotes for Quoted.

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+  | frmt == Format "xwiki" = return $ pack str
+  | otherwise              = "" <$ report (InlineNotRendered il)
+
+-- TODO: Handle anchors
```
It would be good to handle internal links.

dchenbecker replied:
Can you give me an example of what a `Link` to an internal reference would look like? I don't see anything in the testsuite.native file.

mb21 replied:
The following markdown should result in a sensible xwiki representation with `pandoc input.md -t xwiki`

```
# title {#myid}

some internal link that [links to the tile](#myid)
```

dchenbecker replied:
Test produces reasonable output, but only if the user adds the "id" extension:

https://extensions.xwiki.org/xwiki/bin/view/Extension/Id%20Macro

There's no default support for arbitrary anchors in xwiki.

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+    params = intercalate " " [
+        if Text.null alt' then "" else "alt=\"" <> alt' <> "\"",
+          if Text.null titText then "" else "title=\"" <> titText <> "\""
+        ]
```
I don't think this is quite what you want.  If alt is null and titText is "hi", for example, you'll get `intercalate " " ["", "hi"]` which has a leading space. If they're both null, you'll get a string with a space.

jgm replied:
I think a better approach would be
```
params = mconcat [ if Text.null alt' then "" else " alt=\""...
```
That is, just include the space at the beginning of the attribute if it's not null. That saves you an intercalate and a filter.

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+
+inlineToXWiki (Note contents) = blockListToXWiki contents
+
+-- FIXME: support attrs
```
In particular, it would be good to support anchors.

dchenbecker replied:
I think I have this fixed, but there don't appear to be any Spans at all in testsuite.native

dchenbecker replied:
Tested with a simple input html:

```html
<span id="foo">This is fooey</span>
```

produces:

```xwiki
{{ id name="foo" /}}This is fooey
```

Again, dependent on https://extensions.xwiki.org/xwiki/bin/view/Extension/Id%20Macro

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+        ]
+  return $ "[[image:" <> (pack source) <> (if Text.null params then "" else "|| " <> params) <> "]]"
+
+inlineToXWiki (Note contents) = blockListToXWiki contents
```
Is there a more graceful fallback?  (Maybe just printing `[`...`]` around the note's contents would help identify it as a note.)

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+  return $ vcat contents' <> if Text.null lev then "\n" else ""
+  
+
+listItemToXWiki :: PandocMonad m => [Block] -> XWikiReader m Text
```
If this doesn't handle loose lists (you say numbering restarts), then one solution might be to strip trailing newlines from the items, so they don't end up with blank lines in them even if paragraphs are used. This will still break when the items have multiple blocks in them, of course, but it would be a lot better in common cases.

dchenbecker replied:
Thanks, that seems to work better.

jgm commented on src/Text/Pandoc/Writers/XWiki.hs:
```diff
@@ -0,0 +1,256 @@
+  return $ marker <> " " <> labelText <> "\n" <>
+    intercalate "\n" (map (\d -> (Data.Text.init marker) <> ": " <> d) contents)
+
+escapeXWikiString :: Text -> Text
```
Is this really all that needs escaping?
```
What if I write a dash--like this--and I don't want emphasis?
```

dchenbecker replied:
Are you saying I should escape things like "--", "**", and "//"?

dchenbecker replied:
I think I have this fixed, but there aren't any examples like this in testsuite.native

dchenbecker replied:
Had to make a minor fix for this.

Tested with simple HTML input:

```html
<p>This--and not this--is ok.</p>
<p>So **is this**?</p>
<p>And what //about this//?</p>
```

produces:

```xwiki
This~-~-and not this~-~-is ok.

So ~*~*is this~*~*?

And what ~/~/about this~/~/?
```

jgm commented on test/writer.xwiki:
```diff
@@ -0,0 +1,862 @@
+
+Here’s a regular paragraph.
+
+
```
There are two blank lines between each paragraph.  Is that needed in xwiki?  If not it should be avoided.

Note that Text.Pandoc.Pretty might be worth using instead of just generating Text.

jgm commented on test/writer.xwiki:
```diff
@@ -0,0 +1,862 @@
+> Nested block quotes:
+> 
+> 
+> > nested
```
Does this style of nested block quote work in xwiki?

jgm commented on test/writer.xwiki:
```diff
@@ -0,0 +1,862 @@
+
+Hr’s:
+
+
```
Why all the blank lines here?

dchenbecker replied:
These are RawBlocks containing HTML markup. I tried using the `html` macro, which works for some items, but for most items in testsuite.native I get:

```
Failed to execute the [html] macro. Cause: [When using the HTML macro inline, you can only use inline HTML content. Block HTML content (such as tables) cannot be displayed.
```

Previously I was reporting `BlockNotRendered`, which seems like it's probably the appropriate response.

jgm commented:
I've added a bunch of comments.  I don't think this is ready to merge into master yet, but it's getting a lot closer!

dchenbecker commented:
Appreciate all of the comments! I'll see what I can do to start tackling them.

tarleb commented:
@dchenbecker this seems to be related to 7e389cb3dbdc11126b9bdb6a7741a65e5a94a43e. Try adding `{-# LANGUAGE NoImplicitPrelude #-}` to the top of the file, then import prelude manually:

``` haskell
module Text.Pandoc.Writers.XWiki ( writeXWiki ) where
import Prelude
```

dchenbecker commented:
I've addressed most of the top-level comments, but I need some guidance on cases that aren't currently covered in the `testsuite.native` file.

kevinszatkowski commented:
@dchenbecker, Thanks for taking this on, I needed to convert media wiki to xwiki and your addition has been a lifesaver.  I did run into an issue with tables using the attached
[test-table.txt](https://github.com/jgm/pandoc/files/2810677/test-table.txt) as input the output didn't format correctly (not super hard to fix by hand).  Everything else so far has worked like a champ, thanks again!

dchenbecker commented:
I think I addressed most of the issues other than the raw HTML and the table formatting issue that @kevinszatkowski submitted. I got a weird build error, so I rebased onto current master, and I'll try to take a stab at fixing the table formatting.

dchenbecker commented:
> @dchenbecker, Thanks for taking this on, I needed to convert media wiki to xwiki and your addition has been a lifesaver. I did run into an issue with tables using the attached
> [test-table.txt](https://github.com/jgm/pandoc/files/2810677/test-table.txt) as input example the output didn't format correctly (not super hard to fix by hand). Everything else so far has worked like a champ, thanks again!

@kevinszatkowski  I switched to using "\\" for hard breaks, which I think fixes the table formatting issue. If you have a moment, could you test it out?

dchenbecker commented:
I have no idea what's going on with the most recent build failures. I haven't touched that code, and at least GHC 8.6 is passing. I tried rebasing onto master to see if I was maybe missing a commit, but that didn't help.

mb21 commented:
I think the last commit on current master broke the tests...

jgm commented:
It's an update to cmark-gfm-hs that broke things.

Derek Chen-Becker <notifications@github.com> writes:

> I have no idea what's going on with the most recent build failures. I haven't touched that code, and at least GHC 8.6 is passing. I tried rebasing onto master to see if I was maybe missing a commit, but that didn't help.
>
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/jgm/pandoc/pull/4167#issuecomment-472426386

dchenbecker commented:
OK, build passes and all issues are addressed, other than raw HTML, which I don't see a solution for because XWiki doesn't support it.

jgm commented:
This looks good!  If you want to tweak the code later on, feel free to submit additional PRs.  But this is definitely worth adding in its current state.

dchenbecker commented:
Awesome, I'll keep tweaking, and I'll start thinking about what it would take for reader support :)

geofflangenderfer commented:
@dchenbecker @jgm do y'all think it would be tough to write a xwiki -> markdown parser?

dchenbecker commented:
@geofflangenderfer I'll have to look at what's involved in writing the parser side of things, but the writer wasn't as hard as I initially thought it would be. I'll be upfront: between work and kids it's going to be slow going

geofflangenderfer commented:
I may give it a shot if I'm feeling ambitious :)
